### PR TITLE
[MGR+1] Change, document `LOG_UNSERIALIZABLE_REQUESTS`

### LINKED ISSUE #1592: Document LOG_UNSERIALIZABLE_REQUESTS setting
It's implemented in [scrapy/core/scheduler.py#L31](https://github.com/scrapy/scrapy/blob/master/scrapy/core/scheduler.py#L31) and not mentioned anywhere in docs. I think it's really useful, could be mentioned at least in http://scrapy.readthedocs.org/en/latest/topics/jobs.html#request-serialization.

Since it isn't documented we could take the chance and rename it as well, this kind of settings are usually named `<something>_DEBUG`.

### PR DESCRIPTION
**Pull request by darshanime**
Changed `LOG_UNSERIALIZABLE_REQUESTS` to `SCHEDULER_DEBUG`. Document it in the `docs`. Changed behavior to log once and collect stats.
Fixes #1592

### CODE REVIEW

**Comment by darshanime** [comment]
Kindly review @curita, @nramirezuy. 
I need to write tests yet. Where should they go ? (I am thinking about `test.util_reqser.py`)

File: scrapy/core/scheduler.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def from_crawler(cls, crawler):
         dupefilter = dupefilter_cls.from_settings(settings)
         dqclass = load_object(settings['SCHEDULER_DISK_QUEUE'])
         mqclass = load_object(settings['SCHEDULER_MEMORY_QUEUE'])
-        logunser = settings.getbool('LOG_UNSERIALIZABLE_REQUESTS')
```
Comment (nramirezuy) [inline scrapy/core/scheduler.py]: We never completely remove older cases. This settings should still remain in the code with a deprecation warning.

Comment (darshanime) [inline scrapy/core/scheduler.py]: Made the change in the latest commit

**Comment by darshanime** [comment]
@nramirezuy, can you please review the PR? Thanks.

**Comment by lopuhin** [comment]
Anything blocking the merge here?
I have no opinion on the best name for the setting, but at least mentioning the setting in the docs could save a lot of debugging time (as the serialization error could be raised in several different places).

**Comment by darshanime** [comment]
Can we merge this PR? ping @curita

File: docs/topics/jobs.rst [hunk 2]
```diff
@@ -96,4 +96,6 @@ But this will::
         somearg = response.meta['somearg']
         print "the argument passed is:", somearg
 
+If you wish to log the requests that couldn't be serialized, you can set the ``SCHEDULER_DEBUG`` setting to ``True`` in the project's settings page. It is ``False`` by default.
```
Comment (redapple) [inline docs/topics/jobs.rst]: Could you wrap the line around 80 chars?

Comment (darshanime) [inline docs/topics/jobs.rst]: okay. Also, I'll pull in the latest code and merge the conflict. Is there anything else?

Comment (darshanime) [inline docs/topics/jobs.rst]: okay. Also, I'll pull in the latest code and resolve the conflict. Is there anything else?

File: docs/topics/settings.rst [hunk 3]
```diff
@@ -897,6 +897,16 @@ Default: ``'scrapy.core.scheduler.Scheduler'``
...
 The scheduler to use for crawling.
 
+.. setting:: SCHEDULER_DEBUG
+
+SCHEDULER_DEBUG
+---------------
+
+Default: ``False``
+
+Setting to ``True`` will log the first unserializable request encountered.
```
Comment (redapple) [inline docs/topics/settings.rst]: `SCHEDULER_DEBUG` reads like "enable to help debugging the scheduler component."
Wording it as "it's only to log unserializable requests" feels a bit limiting.

What about something along the lines of
> enable this setting to log debug information about Scrapy requests scheduler.
> This currently logs if requests cannot be serialized to disk. If this happens, a line will appear in the logs (only once), and a `scheduler/unserializable` stats counter tracks how many times this happened.

Including an example log line would also be good.

**Comment by darshanime** [comment]
@redapple kindly review!

File: docs/topics/settings.rst [hunk 4]
```diff
@@ -1026,6 +1026,23 @@ Default: ``'scrapy.core.scheduler.Scheduler'``
...
 
+.. setting:: SCHEDULER_DEBUG
+
+SCHEDULER_DEBUG
+---------------
+
+Default: ``False``
+
+Setting to ``True`` will log debug information about the requests scheduler.
+This currently logs(only once) if the requests cannot be serialized to disk.
```
Comment (redapple) [inline docs/topics/settings.rst]: missing space: `logs (only once)`

File: scrapy/core/scheduler.py [hunk 5]
```diff
@@ -84,11 +84,16 @@ def _dqpush(self, request):
...
             reqd = request_to_dict(request, self.spider)
             self.dqs.push(reqd, -request.priority)
-        except ValueError as e: # non serializable request
+        except ValueError as e:  # non serializable request
             if self.logunser:
-                logger.error("Unable to serialize request: %(request)s - reason: %(reason)s",
-                             {'request': request, 'reason': e},
+                msg = ("Unable to serialize request: %(request)s - reason:"
+                       " %(reason)s - no more unserializable requests will be"
+                       " logged (stats being collected)")
```
Comment (redapple) [inline scrapy/core/scheduler.py]: `(stats being collected)` could be changed to something like `(see 'scheduler/unserializable' stats counter)`

File: docs/topics/jobs.rst [hunk 6]
```diff
@@ -96,4 +96,8 @@ But this will::
         somearg = response.meta['somearg']
         print "the argument passed is:", somearg
 
+If you wish to log the requests that couldn't be serialized, you can set the
+``SCHEDULER_DEBUG`` setting to ``True`` in the project's settings page.
```
Comment (redapple) [inline docs/topics/jobs.rst]: You can add a direct link to the setting:

    you can set the :setting:`SCHEDULER_DEBUG` to ...

File: docs/topics/settings.rst [hunk 7]
```diff
@@ -1026,6 +1026,23 @@ Default: ``'scrapy.core.scheduler.Scheduler'``
...
+Stats counter (``scheduler/unserializable``) tracks the number of times this happens.
+
+Example entry in logs::
+
+    1956-01-31 00:00:00+0800 [scrapy] ERROR: Unable to serialize request: <request>
```
Comment (redapple) [inline docs/topics/settings.rst]: Can you use complete log line with request and exception, eventually obfuscating domains?

**Comment by redapple** [comment]
LGTM

**Comment by darshanime** [comment]
@redapple, can we merge this?

**Comment by redapple** [comment]
Fine by me. @kmike , @eliasdorneles , @dangra , what do you think?

**Comment by dangra** [comment]
LGTM.

**Comment by eliasdorneles** [comment]
Looks good, merging -- thanks @darshanime !

**Pull request merged**